mailbox/elasticsearch-v6/src/main/java/org/apache/james/mailbox/elasticsearch/v6/query/CriterionConverter.java

```diff
@@ -0,0 +1,313 @@
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.stream.Collector;
+import java.util.stream.Stream;
+
+import javax.mail.Flags;
+
+import org.apache.james.mailbox.elasticsearch.v6.json.HeaderCollection;
+import org.apache.james.mailbox.elasticsearch.v6.json.JsonMessageConstants;
+import org.apache.james.mailbox.model.SearchQuery;
+import org.apache.james.mailbox.model.SearchQuery.Criterion;
+import org.apache.james.mailbox.model.SearchQuery.HeaderOperator;
+import org.apache.lucene.search.join.ScoreMode;
+import org.elasticsearch.index.query.BoolQueryBuilder;
+import org.elasticsearch.index.query.QueryBuilder;
+import org.elasticsearch.index.query.QueryBuilders;
+
+public class CriterionConverter {
+
+    private final Map<Class<?>, Function<SearchQuery.Criterion, QueryBuilder>> criterionConverterMap;
+    private final Map<Class<?>, BiFunction<String, SearchQuery.HeaderOperator, QueryBuilder>> headerOperatorConverterMap;
+
+    public CriterionConverter() {
+        criterionConverterMap = new HashMap<>();
+        headerOperatorConverterMap = new HashMap<>();
+        
+        registerCriterionConverters();
+        registerHeaderOperatorConverters();
+    }
+
+    private void registerCriterionConverters() {
+        registerCriterionConverter(SearchQuery.FlagCriterion.class, this::convertFlag);
+        registerCriterionConverter(SearchQuery.UidCriterion.class, this::convertUid);
+        registerCriterionConverter(SearchQuery.ConjunctionCriterion.class, this::convertConjunction);
+        registerCriterionConverter(SearchQuery.HeaderCriterion.class, this::convertHeader);
+        registerCriterionConverter(SearchQuery.TextCriterion.class, this::convertTextCriterion);
+        registerCriterionConverter(SearchQuery.CustomFlagCriterion.class, this::convertCustomFlagCriterion);
+        
+        registerCriterionConverter(SearchQuery.AllCriterion.class,
+            criterion -> matchAllQuery());
+        
+        registerCriterionConverter(SearchQuery.ModSeqCriterion.class,
+            criterion -> createNumericFilter(JsonMessageConstants.MODSEQ, criterion.getOperator()));
+        
+        registerCriterionConverter(SearchQuery.SizeCriterion.class,
+            criterion -> createNumericFilter(JsonMessageConstants.SIZE, criterion.getOperator()));
+
+        registerCriterionConverter(SearchQuery.InternalDateCriterion.class,
+            criterion -> dateRangeFilter(JsonMessageConstants.DATE, criterion.getOperator()));
+
+        registerCriterionConverter(SearchQuery.AttachmentCriterion.class, this::convertAttachmentCriterion);
+        registerCriterionConverter(SearchQuery.MimeMessageIDCriterion.class, this::convertMimeMessageIDCriterion);
+    }
+    
+    @SuppressWarnings("unchecked")
+    private <T extends Criterion> void registerCriterionConverter(Class<T> type, Function<T, QueryBuilder> f) {
+        criterionConverterMap.put(type, (Function<Criterion, QueryBuilder>) f);
```

Comment:
I prefer that you put the covariance in the field declaration and avoid the cast

Reply:
Well I tried to change it but then it creates an other issue with the `convertCriterion` method that I couldn't really solve. So let's keep it this way for now ok?

Reply:
ok